README.md
> Package captcha provides an easy to use, unopinionated API for captcha generation

<div>

[![GoDoc](https://godoc.org/github.com/steambap/captcha?status.svg)](https://godoc.org/github.com/steambap/captcha)
[![Build Status](https://travis-ci.org/steambap/captcha.svg)](https://travis-ci.org/steambap/captcha)
[![codecov](https://codecov.io/gh/steambap/captcha/branch/master/graph/badge.svg)](https://codecov.io/gh/steambap/captcha)
[![Go Report Card](https://goreportcard.com/badge/github.com/steambap/captcha)](https://goreportcard.com/report/github.com/steambap/captcha)

</div>

## Why another captcha generator?
I want a simple and framework-independent way to generate captcha. It also should be flexible, at least allow me to pick my favorite font.

## install
```
go get github.com/steambap/captcha
```

## usage
```Go
func handle(w http.ResponseWriter, r *http.Request) {
	// create a captcha of 150x50px
	data, _ := captcha.New(150, 50)

	// session come from other library such as gorilla/sessions
	session.Values["captcha"] = data.Text
	session.Save(r, w)
	// send image data to client
	data.WriteImage(w)
}

```

[documentation](https://godoc.org/github.com/steambap/captcha) |
[example](example/basic/main.go)

## sample image
![image](example/captcha.png)

![image](example/captcha-math.png)

## Contributing
If your found a bug, please contribute!
see [contributing.md](contributing.md) for more detail

## License
[MIT](LICENSE.md)


history.md
1.3.0 / 2018-11-7
===================

* Add Palette option

1.2.0 / 2017-12-26
===================

* Add Noise option

1.1.0 / 2017-11-16
===================

* Add WriteJPG and WriteGIF API

1.0.0 / 2017-10-10
===================

* Add LoadFontFromReader API
* Rename WriteTo to WriteImage

0.12.0 / 2017-10-07
===================

* Add FontDPI and FontScale options

0.11.0 / 2017-09-28
===================

* Add NewMathExpr API

0.10.0 / 2017-09-23
===================

* Add LoadFont API

0.9.0 / 2017-09-20
===================

* Initial release

hsva_test.go
package captcha

import (
	"fmt"
	"image/color"
	"testing"
)

func TestHSVAInterface(t *testing.T) {
	var _ color.Color = hsva{}
}

func TestConversionGray(t *testing.T) {
	var white1 color.Color = hsva{h: 0.0, s: 0.0, v: 1.0, a: uint8(255)}
	var white2 color.Color = color.RGBA{R: uint8(255), G: uint8(255), B: uint8(255), A: uint8(255)}

	if err := eq(white1, white2); err != nil {
		t.Fatal(err)
	}

	var black1 color.Color = hsva{h: 0.0, s: 0.0, v: 0.0, a: uint8(255)}
	var black2 color.Color = color.RGBA{R: uint8(0), G: uint8(0), B: uint8(0), A: uint8(255)}

	if err := eq(black1, black2); err != nil {
		t.Fatal(err)
	}

	var gray1 color.Color = hsva{h: 1.0, s: 0.0, v: 0.8, a: uint8(255)}
	var gray2 color.Color = color.RGBA{R: uint8(204), G: uint8(204), B: uint8(204), A: uint8(255)}

	if err := eq(gray1, gray2); err != nil {
		t.Fatal(err)
	}
}

func TestConversionRGB(t *testing.T) {
	var yellow1 color.Color = hsva{h: 60.0 / 360.0, s: 0.6, v: 1.0, a: uint8(255)}
	var yellow2 color.Color = color.RGBA{R: uint8(255), G: uint8(255), B: uint8(102), A: uint8(255)}

	if err := eq(yellow1, yellow2); err != nil {
		t.Fatal(err)
	}

	var green1 color.Color = hsva{h: 120.0 / 360.0, s: 1.0, v: 0.8, a: uint8(255)}
	var green2 color.Color = color.RGBA{R: uint8(0), G: uint8(204), B: uint8(0), A: uint8(255)}

	if err := eq(green1, green2); err != nil {
		t.Fatal(err)
	}

	var teal1 color.Color = hsva{h: 180.0 / 360.0, s: 0.5, v: 0.8, a: uint8(255)}
	var teal2 color.Color = color.RGBA{R: uint8(102), G: uint8(204), B: uint8(204), A: uint8(255)}

	if err := eq(teal1, teal2); err != nil {
		t.Fatal(err)
	}

	var blue1 color.Color = hsva{h: 240.0 / 360.0, s: 0.75, v: 0.8, a: uint8(255)}
	var blue2 color.Color = color.RGBA{R: uint8(51), G: uint8(51), B: uint8(204), A: uint8(255)}

	if err := eq(blue1, blue2); err != nil {
		t.Fatal(err)
	}

	var pink1 color.Color = hsva{h: 300.0 / 360.0, s: 0.2, v: 1.0, a: uint8(255)}
	var pink2 color.Color = color.RGBA{R: uint8(255), G: uint8(204), B: uint8(255), A: uint8(255)}

	if err := eq(pink1, pink2); err != nil {
		t.Fatal(err)
	}
}

func eq(c0, c1 color.Color) error {
	r0, g0, b0, a0 := c0.RGBA()
	r1, g1, b1, a1 := c1.RGBA()
	if r0 != r1 || g0 != g1 || b0 != b1 || a0 != a1 {
		return fmt.Errorf("got  0x%04x 0x%04x 0x%04x 0x%04x\nwant 0x%04x 0x%04x 0x%04x 0x%04x",
			r0, g0, b0, a0, r1, g1, b1, a1)
	}
	return nil
}


captcha_test.go
package captcha

import (
	"bytes"
	"errors"
	"image/color"
	"image/color/palette"
	"image/gif"
	"image/jpeg"
	"math/rand"
	"os"
	"testing"

	"golang.org/x/image/font/gofont/goregular"
)

func TestNewCaptcha(t *testing.T) {
	data, err := New(150, 50)
	if err != nil {
		t.Fatal(err)
	}
	buf := new(bytes.Buffer)
	err = data.WriteImage(buf)
	if err != nil {
		t.Fatal(err)
	}
}

func TestSmallCaptcha(t *testing.T) {
	_, err := New(36, 12)
	if err != nil {
		t.Fatal(err)
	}
}

func TestEncodeJPG(t *testing.T) {
	data, err := New(150, 50)
	if err != nil {
		t.Fatal(err)
	}
	buf := new(bytes.Buffer)
	err = data.WriteJPG(buf, &jpeg.Options{Quality: 70})
	if err != nil {
		t.Fatal(err)
	}
}

func TestEncodeGIF(t *testing.T) {
	data, err := New(150, 50)
	if err != nil {
		t.Fatal(err)
	}
	buf := new(bytes.Buffer)
	err = data.WriteGIF(buf, &gif.Options{})
	if err != nil {
		t.Fatal(err)
	}
}

func TestNewCaptchaOptions(t *testing.T) {
	New(100, 34, func(options *Options) {
		options.BackgroundColor = color.Opaque
		options.CharPreset = "1234567890"
		options.CurveNumber = 0
		options.TextLength = 6
		options.Palette = palette.WebSafe
	})

	NewMathExpr(100, 34, func(options *Options) {
		options.BackgroundColor = color.Black
	})
}

func TestNewMathExpr(t *testing.T) {
	_, err := NewMathExpr(150, 50)
	if err != nil {
		t.Fatal(err)
	}
}

func TestCovNilFontError(t *testing.T) {
	temp := ttfFont
	ttfFont = nil

	_, err := New(150, 50)
	if err == nil {
		t.Fatal("Expect to get nil font error")
	}

	_, err = NewMathExpr(150, 50)
	if err == nil {
		t.Fatal("Expect to get nil font error")
	}

	ttfFont = temp
}

type errReader struct{}

func (errReader) Read(_ []byte) (int, error) {
	return 0, errors.New("")
}

func TestCovReaderErr(t *testing.T) {
	err := LoadFontFromReader(errReader{})
	if err == nil {
		t.Fatal("Expect to get io.Reader error")
	}
}

func TestLoadFont(t *testing.T) {
	err := LoadFont(goregular.TTF)
	if err != nil {
		t.Fatal("Fail to load go font")
	}

	err = LoadFont([]byte("invalid"))
	if err == nil {
		t.Fatal("LoadFont incorrectly parse an invalid font")
	}
}

func TestLoadFontFromReader(t *testing.T) {
	file, err := os.Open("./fonts/Comismsh.ttf")
	if err != nil {
		t.Fatal("Fail to load test file")
	}

	if err = LoadFontFromReader(file); err != nil {
		t.Fatal("Fail to load font from io.Reader")
	}
}

func TestMaxColor(t *testing.T) {
	var result uint32
	result = maxColor()
	if result != 0 {
		t.Fatalf("Expect max color to be 0, got %v", result)
	}
	result = maxColor(1)
	if result != 1 {
		t.Fatalf("Expect max color to be 1, got %v", result)
	}
	result = maxColor(52428, 65535)
	if result != 255 {
		t.Fatalf("Expect max color to be 255, got %v", result)
	}
	var rng = rand.New(rand.NewSource(0))
	for i := 0; i < 10; i++ {
		result = maxColor(rng.Uint32(), rng.Uint32(), rng.Uint32())
		if result > 255 {
			t.Fatalf("Number out of range: %v", result)
		}
	}
}

func TestMinColor(t *testing.T) {
	var result uint32
	result = minColor()
	if result != 255 {
		t.Fatalf("Expect min color to be 255, got %v", result)
	}
	result = minColor(1)
	if result != 1 {
		t.Fatalf("Expect min color to be 1, got %v", result)
	}
	result = minColor(52428, 65535)
	if result != 204 {
		t.Fatalf("Expect min color to be 1, got %v", result)
	}
	var rng = rand.New(rand.NewSource(0))
	for i := 0; i < 10; i++ {
		result = minColor(rng.Uint32(), rng.Uint32(), rng.Uint32())
		if result > 255 {
			t.Fatalf("Number out of range: %v", result)
		}
	}
}


example/load-font/main.go
package main

import (
	"fmt"
	"html/template"
	"net/http"

	"github.com/steambap/captcha"
	"golang.org/x/image/font/gofont/goregular"
)

func main() {
	err := captcha.LoadFont(goregular.TTF)
	if err != nil {
		panic(err)
	}

	http.HandleFunc("/", indexHandle)
	http.HandleFunc("/captcha", captchaHandle)
	fmt.Println("Server start at port 8080")
	err = http.ListenAndServe(":8080", nil)
	if err != nil {
		panic(err)
	}
}

func indexHandle(w http.ResponseWriter, _ *http.Request) {
	doc, err := template.ParseFiles("index.html")
	if err != nil {
		fmt.Fprint(w, err.Error())
		return
	}
	doc.Execute(w, nil)
}

func captchaHandle(w http.ResponseWriter, _ *http.Request) {
	img, err := captcha.New(150, 50, func(options *captcha.Options) {
		options.FontScale = 0.8
	})
	if err != nil {
		fmt.Fprint(w, nil)
		fmt.Println(err.Error())
		return
	}
	img.WriteImage(w)
}


captcha.go
// Package captcha provides an easy to use, unopinionated API for captcha generation
package captcha

import (
	"bytes"
	"image"
	"image/color"
	"image/draw"
	"image/gif"
	"image/jpeg"
	"image/png"
	"io"
	"math"
	"math/rand"
	"strconv"
	"time"

	"github.com/golang/freetype"
	"github.com/golang/freetype/truetype"
	"golang.org/x/image/font"
)

const charPreset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

var rng = rand.New(rand.NewSource(time.Now().UnixNano()))
var ttfFont *truetype.Font

// Options manage captcha generation details.
type Options struct {
	// BackgroundColor is captcha image's background color.
	// It defaults to color.Transparent.
	BackgroundColor color.Color
	// CharPreset decides what text will be on captcha image.
	// It defaults to digit 0-9 and all English alphabet.
	// ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789
	CharPreset string
	// TextLength is the length of captcha text.
	// It defaults to 4.
	TextLength int
	// CurveNumber is the number of curves to draw on captcha image.
	// It defaults to 2.
	CurveNumber int
	// FontDPI controls DPI (dots per inch) of font.
	// The default is 72.0.
	FontDPI float64
	// FontScale controls the scale of font.
	// The default is 1.0.
	FontScale float64
	// Noise controls the number of noise drawn.
	// A noise dot is drawn for every 28 pixel by default.
	// The default is 1.0.
	Noise float64
	// Palette is the set of colors to chose from
	Palette color.Palette

	width  int
	height int
}

func newDefaultOption(width, height int) *Options {
	return &Options{
		BackgroundColor: color.Transparent,
		CharPreset:      charPreset,
		TextLength:      4,
		CurveNumber:     2,
		FontDPI:         72.0,
		FontScale:       1.0,
		Noise:           1.0,
		Palette:         []color.Color{},
		width:           width,
		height:          height,
	}
}

// SetOption is a function that can be used to modify default options.
type SetOption func(*Options)

// Data is the result of captcha generation.
// It has a `Text` field and a private `img` field that will
// be used in `WriteImage` receiver.
type Data struct {
	// Text is captcha solution.
	Text string

	img *image.NRGBA
}

// WriteImage encodes image data and writes to an io.Writer.
// It returns possible error from PNG encoding.
func (data *Data) WriteImage(w io.Writer) error {
	return png.Encode(w, data.img)
}

// WriteJPG encodes image data in JPEG format and writes to an io.Writer.
// It returns possible error from JPEG encoding.
func (data *Data) WriteJPG(w io.Writer, o *jpeg.Options) error {
	return jpeg.Encode(w, data.img, o)
}

// WriteGIF encodes image data in GIF format and writes to an io.Writer.
// It returns possible error from GIF encoding.
func (data *Data) WriteGIF(w io.Writer, o *gif.Options) error {
	return gif.Encode(w, data.img, o)
}

func init() {
	ttfFont, _ = freetype.ParseFont(ttf)
}

// LoadFont let you load an external font.
func LoadFont(fontData []byte) error {
	var err error
	ttfFont, err = freetype.ParseFont(fontData)
	return err
}

// LoadFontFromReader load an external font from an io.Reader interface.
func LoadFontFromReader(reader io.Reader) error {
	var buf bytes.Buffer
	if _, err := io.Copy(&buf, reader); err != nil {
		return err
	}

	return LoadFont(buf.Bytes())
}

// New creates a new captcha.
// It returns captcha data and any freetype drawing error encountered.
func New(width int, height int, option ...SetOption) (*Data, error) {
	options := newDefaultOption(width, height)
	for _, setOption := range option {
		setOption(options)
	}

	text := randomText(options)
	img := image.NewNRGBA(image.Rect(0, 0, width, height))
	draw.Draw(img, img.Bounds(), &image.Uniform{options.BackgroundColor}, image.ZP, draw.Src)
	drawNoise(img, options)
	drawCurves(img, options)
	err := drawText(text, img, options)
	if err != nil {
		return nil, err
	}

	return &Data{Text: text, img: img}, nil
}

// NewMathExpr creates a new captcha.
// It will generate a image with a math expression like `1 + 2`.
func NewMathExpr(width int, height int, option ...SetOption) (*Data, error) {
	options := newDefaultOption(width, height)
	for _, setOption := range option {
		setOption(options)
	}

	text, equation := randomEquation()
	img := image.NewNRGBA(image.Rect(0, 0, width, height))
	draw.Draw(img, img.Bounds(), &image.Uniform{options.BackgroundColor}, image.ZP, draw.Src)
	drawNoise(img, options)
	drawCurves(img, options)
	err := drawText(equation, img, options)
	if err != nil {
		return nil, err
	}

	return &Data{Text: text, img: img}, nil
}

func randomText(opts *Options) (text string) {
	n := len(opts.CharPreset)
	for i := 0; i < opts.TextLength; i++ {
		text += string(opts.CharPreset[rng.Intn(n)])
	}

	return text
}

func drawNoise(img *image.NRGBA, opts *Options) {
	noiseCount := (opts.width * opts.height) / int(28.0/opts.Noise)
	for i := 0; i < noiseCount; i++ {
		x := rng.Intn(opts.width)
		y := rng.Intn(opts.height)
		img.Set(x, y, randomColor())
	}
}

func randomColor() color.RGBA {
	red := rng.Intn(256)
	green := rng.Intn(256)
	blue := rng.Intn(256)

	return color.RGBA{R: uint8(red), G: uint8(green), B: uint8(blue), A: uint8(255)}
}

func drawCurves(img *image.NRGBA, opts *Options) {
	for i := 0; i < opts.CurveNumber; i++ {
		drawSineCurve(img, opts)
	}
}

// Ideally we want to draw bezier curves
// For now sine curves will do the job
func drawSineCurve(img *image.NRGBA, opts *Options) {
	var xStart, xEnd int
	if opts.width <= 40 {
		xStart, xEnd = 1, opts.width-1
	} else {
		xStart = rng.Intn(opts.width/10) + 1
		xEnd = opts.width - rng.Intn(opts.width/10) - 1
	}
	curveHeight := float64(rng.Intn(opts.height/6) + opts.height/6)
	yStart := rng.Intn(opts.height*2/3) + opts.height/6
	angle := 1.0 + rng.Float64()
	yFlip := 1.0
	if rng.Intn(2) == 0 {
		yFlip = -1.0
	}
	curveColor := randomColorFromOptions(opts)

	for x1 := xStart; x1 <= xEnd; x1++ {
		y := math.Sin(math.Pi*angle*float64(x1)/float64(opts.width)) * curveHeight * yFlip
		img.Set(x1, int(y)+yStart, curveColor)
	}
}

func drawText(text string, img *image.NRGBA, opts *Options) error {
	ctx := freetype.NewContext()
	ctx.SetDPI(opts.FontDPI)
	ctx.SetClip(img.Bounds())
	ctx.SetDst(img)
	ctx.SetHinting(font.HintingFull)
	ctx.SetFont(ttfFont)

	fontSpacing := opts.width / len(text)
	fontOffset := rng.Intn(fontSpacing / 2)

	for idx, char := range text {
		fontScale := 0.8 + rng.Float64()*0.4
		fontSize := float64(opts.height) / fontScale * opts.FontScale
		ctx.SetFontSize(fontSize)
		ctx.SetSrc(image.NewUniform(randomColorFromOptions(opts)))
		x := fontSpacing*idx + fontOffset
		y := opts.height/6 + rng.Intn(opts.height/3) + int(fontSize/2)
		pt := freetype.Pt(x, y)
		if _, err := ctx.DrawString(string(char), pt); err != nil {
			return err
		}
	}

	return nil
}

func randomColorFromOptions(opts *Options) color.Color {
	length := len(opts.Palette)
	if length == 0 {
		return randomInvertColor(opts.BackgroundColor)
	}

	return opts.Palette[rng.Intn(length)]
}

func randomInvertColor(base color.Color) color.Color {
	baseLightness := getLightness(base)
	var value float64
	if baseLightness >= 0.5 {
		value = baseLightness - 0.3 - rng.Float64()*0.2
	} else {
		value = baseLightness + 0.3 + rng.Float64()*0.2
	}
	hue := float64(rng.Intn(361)) / 360
	saturation := 0.6 + rng.Float64()*0.2

	return hsva{h: hue, s: saturation, v: value, a: 255}
}

func getLightness(colour color.Color) float64 {
	r, g, b, a := colour.RGBA()
	// transparent
	if a == 0 {
		return 1.0
	}
	max := maxColor(r, g, b)
	min := minColor(r, g, b)

	l := (float64(max) + float64(min)) / (2 * 255)

	return l
}

func maxColor(numList ...uint32) (max uint32) {
	for _, num := range numList {
		colorVal := num & 255
		if colorVal > max {
			max = colorVal
		}
	}

	return max
}

func minColor(numList ...uint32) (min uint32) {
	min = 255
	for _, num := range numList {
		colorVal := num & 255
		if colorVal < min {
			min = colorVal
		}
	}

	return min
}

func randomEquation() (text string, equation string) {
	left := 1 + rng.Intn(9)
	right := 1 + rng.Intn(9)
	text = strconv.Itoa(left + right)
	equation = strconv.Itoa(left) + "+" + strconv.Itoa(right)

	return text, equation
}
